File: Algorithms/0045_Jump_Game_II/Python/Jump_Game_II_Solution_1.py

# Space: O(n)
# Time: O(n)

class Solution:
    def jump(self, nums):
        length = len(nums)
        if length <= 1: return 0
        cache = [0 for _ in range(length)]

        for i in range(length):
            if nums[i] == 0: continue
            for j in range(i + 1, i + 1 + nums[i]):
                if j >= length: break
                cache[j] = cache[i] + 1 if cache[j] == 0 else cache[j]

        return cache[-1]

File: Algorithms/0045_Jump_Game_II/Java/JumpGameIiSolution1.java

// Space: O(n)
// Time: O(n)

class Solution {
    public int jump(int[] nums) {
        int length = nums.length;
        if (length <= 1) return 0;
        int[] cache = new int[length];
        Arrays.fill(cache, 0);

        for (int i = 0; i < length; i++) {
            if (nums[i] == 0) continue;
            for (int j = i + 1; j < i + 1 + nums[i]; j++) {
                if (j >= length) break;
                if (cache[j] == 0) cache[j] = cache[i] + 1;
            }
        }
        return cache[length - 1];
    }
}




